src/modules/track_model/inc/track_model.h

```diff
@@ -11,29 +11,65 @@
+        virtual types::Error SetRedTrafficLight(const types::BlockId block, const bool on)                     = 0;
+        virtual types::Error SetYellowTrafficLight(const types::BlockId block, const bool on)                  = 0;
+        virtual types::Error SetGreenTrafficLight(const types::BlockId block, const bool on)                   = 0;
+        virtual types::Error SetCommandedSpeed(const types::BlockId block, const types::MetersPerSecond speed) = 0;
+        virtual types::Error SetAuthority(const types::BlockId block, const types::Blocks authority)           = 0;
+        virtual types::Error GetBlockOccupancy(const types::BlockId block, bool &occupied) const               = 0;
+        virtual std::vector<std::vector<types::BlockId>> GetOccupiedTrainBlocks(void)                          = 0;
+};
+
+class SoftwareTrackModel : public TrackModel
+{
+    public:
+        //ADD constructor that takes in a shared pointer to the tick source
+        types::Error SetTrackLayout(const types::TrackId track, const std::vector<types::Block> &blocks);
+        types::TrackId GetTrackId(void);
+        types::Error AddTrainModel(std::shared_ptr<train_model::TrainModel> train);
+        void GetTrainModels(std::vector<std::shared_ptr<train_model::TrainModel>> &trains) const;
+        void Update(void);
+        types::Error SetSwitchState(const types::BlockId block, const bool switched);
+        types::Error SetCrossingState(const types::BlockId block, const bool closed);
+        types::Error SetRedTrafficLight(const types::BlockId block, const bool on);
+        types::Error SetYellowTrafficLight(const types::BlockId block, const bool on);
+        types::Error SetGreenTrafficLight(const types::BlockId block, const bool on);
+        types::Error SetCommandedSpeed(const types::BlockId block, const types::MetersPerSecond speed);
+        types::Error SetAuthority(const types::BlockId block, const types::Blocks authority);
+        types::Error GetBlockOccupancy(const types::BlockId block, bool &occupied) const;
+        types::Error SetBrokenRail(const types::BlockId block, const bool broken);
+        types::Error SetTrackCircuitFailure(const types::BlockId block, const bool track_circuit_failure);
+        types::Error SetPowerFailure(const types::BlockId block, const bool power_failure);
+        types::Error SetExternalTemperature(const types::DegreesFahrenheit temperature);
+        types::Error SetPassengersDeboarding(const types::TrainId train, const uint16_t passengers);
+        types::Block GetBlock(const types::BlockId block);
+        std::vector<std::vector<types::BlockId>> GetOccupiedTrainBlocks(void);
+
+    private:
+        types::TrackId track_;
+        std::vector<types::Block> blocks_;
+        std::vector<std::shared_ptr<train_model::TrainModel>> trains_;
+        std::vector<std::vector<types::BlockId>> occupied_train_blocks_;
```

Comment:
It might make sense to have a struct or some other way to group all the data for each train together rather than trying to keep track of which index in which vector corresponds to which train

Reply:
noted, but will probably avoid doing this unless I need to